presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/SpatialPartitioningInternalAggregateFunction.java

```diff
@@ -0,0 +1,108 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.plugin.geospatial;
+
+import com.esri.core.geometry.Envelope;
+import com.facebook.presto.geospatial.KDBTree;
+import com.facebook.presto.geospatial.Rectangle;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.function.AggregationFunction;
+import com.facebook.presto.spi.function.CombineFunction;
+import com.facebook.presto.spi.function.InputFunction;
+import com.facebook.presto.spi.function.OutputFunction;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.type.StandardTypes;
+import io.airlift.json.JsonCodec;
+import io.airlift.json.JsonCodecFactory;
+import io.airlift.slice.Slice;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static com.facebook.presto.geospatial.KDBTree.buildKDBTree;
+import static com.facebook.presto.geospatial.serde.GeometrySerde.deserializeEnvelope;
+import static com.facebook.presto.plugin.geospatial.GeometryType.GEOMETRY_TYPE_NAME;
+import static com.facebook.presto.spi.type.StandardTypes.INTEGER;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static java.lang.Math.floor;
+import static java.lang.Math.toIntExact;
+
+@AggregationFunction(value = "spatial_partitioning", decomposable = false, isOrderSensitive = true, hidden = true)
+public class SpatialPartitioningInternalAggregateFunction
+{
+    private static final JsonCodec<KDBTree> KDB_TREE_CODEC = new JsonCodecFactory().jsonCodec(KDBTree.class);
+    private static final int MAX_SAMPLE_COUNT = 1_000_000;
+
+    private SpatialPartitioningInternalAggregateFunction() {}
+
+    @InputFunction
+    public static void input(SpatialPartitioningState state, @SqlType(GEOMETRY_TYPE_NAME) Slice slice, @SqlType(INTEGER) long partitionCount)
+    {
+        Envelope envelope = deserializeEnvelope(slice);
+        if (envelope == null) {
+            return;
+        }
+
+        Rectangle extent = new Rectangle(envelope.getXMin(), envelope.getYMin(), envelope.getXMax(), envelope.getYMax());
+
+        if (state.getCount() == 0) {
+            state.setPartitionCount(toIntExact(partitionCount));
+            state.setExtent(extent);
+            state.setSamples(new ArrayList<>());
+        }
+        else {
+            state.setExtent(state.getExtent().merge(extent));
+        }
+
+        List<Rectangle> samples = state.getSamples();
+        if (samples.size() <= MAX_SAMPLE_COUNT) {
+            samples.add(extent);
+        }
+        else {
+            int sampleIndex = (int) floor(ThreadLocalRandom.current().nextDouble() * state.getCount());
+            if (sampleIndex < MAX_SAMPLE_COUNT) {
```

Comment:
Remove this check

Reply:
@arhimondr This check is necessary. Otherwise, we'll be adding *all* the values, not a sample of these.

Reply:
Why `ThreadLocalRandom.current().nextInt(state.getCount())` is not an option?